modules/apps/transfer/keeper/relay_forwarding_test.go

```diff
@@ -612,3 +613,172 @@ Test async ack is properly relayed to middle hop after forwarding transfer compl
+	var total sdk.Coin
+	switch balanceType {
+	case escrow:
+		total = chain.GetSimApp().TransferKeeper.GetTotalEscrowForDenom(chain.GetContext(), denom)
+	case balance:
+		total = chain.GetSimApp().BankKeeper.GetBalance(chain.GetContext(), chain.SenderAccounts[0].SenderAccount.GetAddress(), denom)
+	default:
+		suite.Fail("invalid amountType %s", balanceType)
+	}
+	suite.Require().Equal(amount, total.Amount, errorMsg)
+}
+
+// TestOnTimeoutPacketForwarding tests the scenario in which a packet goes from
+// A to C, using B as a forwarding hop. The packet times out when going to B
+// from C and we verify that funds are properly returned to A.
+func (suite *KeeperTestSuite) TestOnTimeoutPacketForwarding() {
+	pathAtoB, pathBtoC := suite.setUpForwardingPaths()
+
+	amount := sdkmath.NewInt(100)
+	coin := sdk.NewCoin(sdk.DefaultBondDenom, amount)
+	sender := suite.chainA.SenderAccounts[0].SenderAccount
+	receiver := suite.chainC.SenderAccounts[0].SenderAccount
+
+	denomA := types.NewDenom(coin.Denom)
+	denomAB := types.NewDenom(coin.Denom, types.NewTrace(pathAtoB.EndpointB.ChannelConfig.PortID, pathAtoB.EndpointB.ChannelID))
+	denomABC := types.NewDenom(coin.Denom, append(denomAB.Trace, types.NewTrace(pathBtoC.EndpointB.ChannelConfig.PortID, pathBtoC.EndpointB.ChannelID))...)
+
+	originalABalance := suite.chainA.GetSimApp().BankKeeper.GetBalance(suite.chainA.GetContext(), sender.GetAddress(), coin.Denom)
+
+	forwarding := &types.Forwarding{
+		Hops: []types.Hop{
+			{
+				PortId:    pathBtoC.EndpointA.ChannelConfig.PortID,
+				ChannelId: pathBtoC.EndpointA.ChannelID,
+			},
+		},
+	}
+
+	transferMsg := types.NewMsgTransfer(
+		pathAtoB.EndpointA.ChannelConfig.PortID,
+		pathAtoB.EndpointA.ChannelID,
+		sdk.NewCoins(coin),
+		sender.GetAddress().String(),
+		receiver.GetAddress().String(),
+		suite.chainA.GetTimeoutHeight(),
+		0, "",
+		forwarding,
+	)
+
+	result, err := suite.chainA.SendMsgs(transferMsg)
+	suite.Require().NoError(err) // message committed
+
+	// parse the packet from result events and recv packet on chainB
+	packet, err := ibctesting.ParsePacketFromEvents(result.Events)
+	suite.Require().NoError(err)
+	suite.Require().NotNil(packet)
+
+	err = pathAtoB.EndpointB.UpdateClient()
+	suite.Require().NoError(err)
+
+	// Receive packet on B.
+	result, err = pathAtoB.EndpointB.RecvPacketWithResult(packet)
+	suite.Require().NoError(err)
+	suite.Require().NotNil(result)
+
+	err = pathBtoC.EndpointA.UpdateClient()
+	suite.Require().NoError(err)
+
+	err = pathBtoC.EndpointB.UpdateClient()
+	suite.Require().NoError(err)
+
+	// Make sure funds went from A to B's escrow account.
+	suite.assertAmountOnChain(suite.chainA, balance, originalABalance.Amount.Sub(amount), denomA.IBCDenom(), "Chain A should have less funds")
+	suite.assertAmountOnChain(suite.chainB, escrow, amount, denomAB.IBCDenom(), "Chain B's escrow should have coins")
+
```

Comment:
You could also verify the forwarded packet exists in state using `GetForwardedPacket`. And once https://github.com/cosmos/ibc-go/issues/6546 is done, then you can verify after the timeout executes on B that the forwarded packet has been deleted.

Reply:
Done!